### PR TITLE
[FIX] base,hr_timesheet: uninstall project's 'timesheet' config

### PR DESCRIPTION
**Pull request by auon-odoo**
Steps :
Go to project > settings.
Check 'timesheet' config and save.
Uncheck it back and save.

Issue :
See errors in the logs, due to on delete restrict.

Cause :
Record 'hr_timesheet.internal_project_default_stage' from data
is deleted when hr_timesheet is uninstalled.
Yet some project task records still use it, so we can't delete it.

Fix :
We can't delete the tasks because some timesheets might be linked to them.
At uninstall, copy the stage and move the tasks in the copy to allow its deletion.

task-2998416

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by auon-odoo** [comment]
@Feyensv, I'll need your review on this please. Have a nice end of your time off!

**Comment by Feyensv** [comment]
Well, don't have any strong opposition to it, if @rco-odoo is ok to backport it in 15.0 :thinking:

**Comment by Feyensv** [comment]
Also
```
Then 'industry_fsm_sale' is in to_install (??),
so it is installed back, which makes no difference
with the original state.
```

is quite unclear, probably needs a better explanation/investigation of the problem (cause i don't really get how my changes fix it).

File: odoo/addons/base/models/res_config.py [hunk 1]
```diff
@@ -20,27 +20,14 @@ class ResConfigModuleInstallationMixin(object):
     def _install_modules(self, modules):
         """ Install the requested modules.
 
-        :param modules: a list of tuples (module_name, module_record)
+        :param modules: a recordset of ir.module.module records
```
Comment (IT-Ideas) [inline odoo/addons/base/models/res_config.py:23]: This is an issue in stable as this changes the signature of the def.

**Review by IT-Ideas — COMMENTED** [review]
Hello @auon-odoo 👋🏼 
Thanks for your work 🙂 

I however don't feel confortable with base and the things you are modifying in it.
@rco-odoo , could you have a look at the problem Audric is addressing and advising on the implementation?

Thanks!

File: addons/hr_timesheet/__init__.py [hunk 2]
```diff
@@ -48,3 +48,8 @@ def _uninstall_hook(cr, registry):
     project_ids = env['res.company'].search([('internal_project_id', '!=', False)]).mapped('internal_project_id')
     if project_ids:
         project_ids.write({'active': False})
+
+    stage_copy = env.ref('hr_timesheet.internal_project_default_stage').copy()
+    env['project.task'].with_context(active_test=False).search(
+        [('stage_id', '=', env.ref('hr_timesheet.internal_project_default_stage').id)]
+    ).stage_id = stage_copy
```
Comment (rco-odoo) [inline addons/hr_timesheet/__init__.py:55]: Copying looks like a bad idea to me.
Why don't you simply delete the corresponding `ir.model.data` record?  When deleted, the existing stage remains, but is no longer "recognized" as `hr_timesheet.internal_project_default_stage`...

File: odoo/addons/base/models/res_config.py [hunk 3]
```diff
@@ -472,8 +467,7 @@ def _get_classified_fields(self):
             elif name.startswith('module_'):
                 if field.type not in ('boolean', 'selection'):
                     raise Exception("Field %s must have type 'boolean' or 'selection'" % field)
-                module = IrModule.sudo().search([('name', '=', name[7:])], limit=1)
-                modules.append((name, module))
+                modules += IrModule.sudo().search([('name', '=', name[7:])], limit=1)
```
Comment (rco-odoo) [inline odoo/addons/base/models/res_config.py:470]: We could get the same effect by collecting all module names, search for them once, and then rebuild the list as before.  Something like:
```py
module_names = []
for name, field in self._fields.items():
    ...
    elif name.startswith('module_'):
        ...
        module_names.append(name[7:])
    ...

name_modules = {
    module.name: module
    for module in IrModule.sudo().search([('name', 'in', module_names)])
}
modules = [
    ('module_' + name, name_modules.get(name, IrModule)
    for name in module_names
]
```
and everything else remains the same.  All the records that appear in the list `modules` come from the same recordset, so they will be read as a single batch.

File: addons/hr_timesheet/__init__.py (hunk 2 above)
Comment (auon-odoo) [reply addons/hr_timesheet/__init__.py]: Oh, that's pretty cool! Thank you :)

**Comment by IT-Ideas** [comment]
@robodoo rebase-ff

**Pull request closed without merging**